build/compile_descriptor

```diff
@@ -1,62 +1,66 @@
-#! /bin/sh
+#!/usr/bin/env escript
+%% -*- erlang -*-
+-include_lib("kernel/include/file.hrl").
 
-# This is a simple form of make, called as a post_hook from rebar,
-# but depending on make for rebar support seemed ... wrong somehow
 
-is_up_to_date () {
-    target="$1"
-    shift
-    test ! -f "$target" && return 1
-    for f in "$@"
-    do
-        test "$f" -nt "$target" && return 1
-    done
-    return 0
-}
+is_up_to_date([Target|CheckResources]) ->
+    case file:read_file_info(Target) of
+        {ok, #file_info{mtime = TargetMTime}} ->
+            lists:all(fun(File) ->
+                case file:read_file_info(File) of
+                    {ok, #file_info{mtime = MTime}} when MTime =< TargetMTime ->
+                        true;
+                    _ ->
+                        false
+                end
+                      end, CheckResources);
+        _ ->
+            false
+    end.
 
-# The || exit $? is needed, if sh is a bash, then with some
-# versions of bash with -e, execution does not terminte if a
-# subshell exits with non-zero exit code
+check_source(FileList, EchoString, Cmd) ->
+    case is_up_to_date(FileList) of
+        true ->
+            ok;
+        _ ->
+            io:format("~s~n",[EchoString]),
+            os:cmd(Cmd)
+    end.
 
-set -e
-
-if [ -d "$REBAR_DEPS_DIR/gpb" ]; then
-    # rebar3
-    EBIN="$REBAR_DEPS_DIR/gpb/ebin"
-    TEST="$REBAR_DEPS_DIR/gpb/test"
-    mkdir -p "$EBIN"
-else
-    # rebar2
-    EBIN="./ebin"
-    TEST="./test"
-fi
-
-is_up_to_date descr_src/gpb_descriptor.erl \
-          priv/proto3/google/protobuf/descriptor.proto || \
-    ( echo "Compiling descriptor.proto..."
-    erl +B -noinput -pa "$EBIN" \
-        -I priv/proto3/google/protobuf -o descr_src \
-    -modprefix gpb_ \
-        -s gpb_compile c descriptor.proto ) || exit $?
-is_up_to_date "$EBIN/gpb_descriptor.beam" descr_src/gpb_descriptor.erl || \
-    ( echo "Compiling gpb_descriptor.erl..."
-    erlc -Idescr_src -Iinclude -o "$EBIN" +debug_info \
-        descr_src/gpb_descriptor.erl ) || exit $?
-is_up_to_date "$EBIN/gpb_compile_descr.beam" \
-    descr_src/gpb_compile_descr.erl \
-    descr_src/gpb_descriptor.hrl || \
-    ( echo "Compiling gpb_compile_descr.erl..."
-    erlc -Idescr_src -Iinclude -o "$EBIN" +debug_info \
-    descr_src/gpb_compile_descr.erl ) || exit $?
-if [ -d "$TEST" ]; then
-    # eg: "rebar3 eunit" or "rebar2 eunit"
-    # but not "rebar3 compile"
-    # and not when compiling gpb as a dependency, since
-    # the test directory does not exist in the hex package.
```

Comment:
Please readd relevant parts of the comment to the new code, as it provides explanations for why things look the way it does.